crates/sui-adapter/src/adapter.rs

```diff
@@ -1100,6 +1105,146 @@ pub fn resolve_and_type_check(
     })
 }
 
+// Validates a primitive argument
+fn validate_primitive_arg(
```

Comment:
Correct me if I'm wrong, but I feel like a lot of the complications for this code are a result of not having a full TypeLayout for the type, which is likely because we don't have a VM session at this point.

So when trying to think about simplifying this code path, perhaps we should just make function to create MoveTypeLayouts for our known types?
We would just have:
- UTF8
- ASCII
- ID
  - I will need UID later, more on that in a bit
- T -> Option<T>

Then we can write a visitor for the MoveValue, something like
`visit_structs(value: MoveValue, should_visit: Fn(&StructTag) -> bool, visit: impl FnMut(&MoveStruct) -> ())`
We can then write the UTF8 and ASCII validation with the visitor. 
We could also just have `should_visit` as a single `target_tag: StructTag` and then call `visit_structs` twice, which might be easier.

I'm advocating for this approach because I actually _might_ to need this `visit_structs` nonsense for dealing with wrapped objects in the dynamic child world. (Me needing it will depend on some factors we decide on with regards to inconsistent child loading and shared object creation)

Regardless, I think this approach will be drastically simpler than what is here, and hopefully not too different in performance

Reply:
I think what you propose would be more general but I would argue that it wouldn't be simpler. In my head, the current implementation is a specialized version of what you propose, targeting UTF8 and ASCII strings.

I am not sure how you can recover type layout without knowing what value you are dealing with, unless you would like to have a general conversion between `SignatureToken` and `MoveTypeLayout` (at least part of the reason we could not use VM's layout creation support is that we don't have `TypeTag` or `StructTag` but only `SignatureToken`).

If we want to support creation of `MoveTypeLayout` for our know types, we'd need something like my `string_arg` function to determine the type we are dealing with.

The `string_arg_move_layout` function is indeed the one that creates the `MoveTypeLayout` for ASCII and UTF8 strings. 

The `validate_string_move_value` function is, in a sense, a visitor specialized towards a specific (expected) `MoveTypeLayout` and the `validate_string` function is the equivalent of the `visit` closure (and I was considering using it as such but decided against it given that we are only dealing with two different `MoveTypeLaout`s here).

If there is a reason to generalize it, it's probably doable (though again, I am not sure how to more elegantly specialize it towards our know types only), but I did not see a good reason for it then, and I am not sure if are convinced even now that this would be needed.

As always, I may be missing a ton of subtlety here, so happy to discuss this further!

Reply:
All that being said, when adding support for this in Sui JSON, I removed some of the code that might have looked a bit too ad hoc (as it was already handled in Sui JSON code). The Sui JSON support does not quite work yet, but I wanted to push the changes either way to help with the discussion (as some of the code needed to move to Sui JSON from the adapter to support UTF8 and ASCII strings at that level).

Reply:
> unless you would like to have a general conversion between SignatureToken and MoveTypeLayout

We are basically already doing this, for `is_primitive`. I just overly optimized that function to avoid creating a `TypeTag`. But we could just do `primitive_type(view: &BinaryIndexedView, function_type_arguments: &[TypeTag], t: &SignatureToken) -> Option<MoveTypeLayout>` to replace that function, where it returns Some iff is_primitive is true. It should simplify this logic a bit too since you don't need separate functions to traverse the type

But again, I will need likely this `visit_structs`, so I'm trying to imagine writing this code to verify strings where `visit_structs` already exists. 
It is possible that we don't need `visit_structs`, but it at least feels a bit more principled to write this code in that manner.

Reply:
> > unless you would like to have a general conversion between SignatureToken and MoveTypeLayout
> 
> We are basically already doing this, for `is_primitive`. I just overly optimized that function to avoid creating a `TypeTag`. But we could just do `primitive_type(view: &BinaryIndexedView, function_type_arguments: &[TypeTag], t: &SignatureToken) -> Option<MoveTypeLayout>` to replace that function, where it returns Some iff is_primitive is true. It should simplify this logic a bit too since you don't need separate functions to traverse the type

Good idea! I have "merged" the two functions we had for checking primitive value and for creating the layout.

> But again, I will need likely this `visit_structs`, so I'm trying to imagine writing this code to verify strings where `visit_structs` already exists. It is possible that we don't need `visit_structs`, but it at least feels a bit more principled to write this code in that manner.

I am really not sure if a general-purpose visitor is the right solution here. After all we are not validating a general MoveLayout but a very specific one - a nested vector of structs. I am totally fine implementing it, though, if I know what the other use is - without it I am left guessing how it should behave, what arguments to take, what values to return etc.